pkg/regionpicker/cache.go

```diff
@@ -0,0 +1,119 @@
+package regionpicker
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	// RegionCachePath is the $HOME/<RegionCachePath> location of the regionpicker cache
+	RegionCachePath = filepath.Join(".outreach", ".cache", "box")
+	// RegionCacheFile is the name of the regionpicker cache file
+	RegionCacheFile = "regionpicker.json"
+)
+
+// cacheEntry is information on a cloud's cache entry
+type cacheEntry struct {
+	// LastBest was the last best region that was determined.
+	LastBest RegionName `json:"last_best"`
+
+	// LastUpdatedAt is the last time this file was updated. UTC.
+	LastUpdatedAt time.Time `json:"last_updated_at"`
+}
+
+type cacheFile struct {
+	// Clouds is a map of cloudName to cache entries
+	Clouds map[CloudName]*cacheEntry
+}
+
+// loadCache reads the cache from disk
+func loadCache() (*cacheFile, error) {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get user homedir")
+	}
+
+	f, err := os.Open(filepath.Join(homeDir, RegionCacheFile, RegionCacheFile))
+	if err != nil {
+		return nil, err
+	}
+
+	var cache cacheFile
+	if err := json.NewDecoder(f).Decode(&cache); err != nil {
+		return nil, errors.Wrap(err, "failed to decode cache")
+	}
+
+	return &cache, nil
+}
+
+// saveCache saves the provided cacheFile to disk
+func saveCache(c *cacheFile) error {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return errors.Wrap(err, "failed to get user homedir")
+	}
+
+	cacheDir := filepath.Join(homeDir, RegionCacheFile)
+	if err := os.MkdirAll(cacheDir, 0755); err != nil {
+		return errors.Wrap(err, "failed to create cache dir")
+	}
+
+	f, err := os.Create(filepath.Join(cacheDir, RegionCacheFile))
+	if err != nil {
+		return errors.Wrap(err, "failed to create cache file")
+	}
+
+	return json.NewEncoder(f).Encode(c)
+}
+
+// FindCachedBest returns the best region but uses a cache first.
+func FindCachedBest(cloud CloudName, allowedRegions []RegionName, logger logrus.FieldLogger) (RegionName, error) {
```

Comment:
This is an odd API -- why do we need AllowedRegions here? It makes it questionable to write this to a cache as the cache can only be used if the allowedRegions matched the value used originally.

Also, what you are trying to do is a ReadThrough cache for which there is a nice pattern [here](https://github.com/getoutreach/datastores/blob/main/pkg/cache/readthrough.go#L16) (you'd have to implement a cache interface that goes directly against the network and then you can automatically combine that with a cache.File via cache.ReadThrough).

Reply:
This is entirely optional but if you went down the route I suggested with cache.Cache interface implementations, there is a very clean way to log cache hits and misses - by wrapping the network call and the cache storage separately (reads from network imply cache miss, writes to storage imply successful fetch etc).

Reply:
Thinking about it a bit more, I think a much better option is to cache all the `ping` durations for the regions rather than `last best time`.  With that approach, you are pushing the cache way down (i.e. your call to `ping` would use the cache and return it if it is there, if it isn't there it would do the actual ping).

The benefit of that approach is that AllowedRegions will not interact with the cache anymore.

Reply:
So the context for "AllowedRegions" is that we need to be able to filter what regions we want to use. For example, we don't want to hit every AZ ever if we're just looking for GCP's multi-region setup `us`, `europe`, and `asia`. 

Feedback on the cache interaction with that is fair, but I'm still somewhat against testing against all regions if we don't need to do so. Will think on this a little bit

Reply:
I think the last comment I had takes care of avoiding pinging regions that were not request, so let me elaborate a bit:

1. Implement pinger:  `Ping(endpoint) (duration time.Duration, time.Expiration)`
2. Implement a cached pinger which reads from the cache if there is an entry for `endpoint` (and it hasn't expired), falling back to the raw pinger above
3. Implement the picker to use this cached pinger.  So, it always does the `best` calculation but this should be fast if the ping durations are in the cache.

Reply:
Ahhhh, I see that makes sense!

Reply:
So, this talks to all the cache stuff. I did this a little differently to allow a global cache in the package (to make it goroutine safe) but also to allow easy of strict typing. If I ever do the work to move datastores/pkg/cache out of that repository then I might re-evaluate, but trying to support that interface without doing that work seemed a little over-optimizing for now and would make this more complex.

PTAL and let me know how you feel about it!

Reply:
Haven't reviewed the implementation in detail yet but the interfaces look fine to me.